### PR TITLE
Truffle optimizations iterators

### PR DESCRIPTION
**Pull request by Alexzerntev**
- Converted libs to nodes from iterators and generators
- Fixed every warning (inlinable nodes, sharing nodes) which reduces memory footprint
- Separated Closures with named arguments execution from unnamed
- Revisited truffle boundaries

### CODE REVIEW

**Comment by miguelbranco80** [comment]
Q: any clear performance improvements?

**Comment by Alexzerntev** [comment]
> Q: any clear performance improvements?

from 12954ms to 11759ms on Pavlos query (on average) (scala is 8000)

(haven't converted lists and aggregations to nodes yet though)

**Comment by Alexzerntev** [comment]
@miguelbranco80 after revisiting TruffleBoundaries and fixing some compilation warnings we are at 6741ms (avg) for Pavlos query (Still work in progress) . Btw https://github.com/raw-labs/snapi/pull/295 broke the PE (fixed in this PR, needed a TruffleBoundary in the RawContext.java

**Comment by Alexzerntev** [comment]
We dropped to 7453ms (avg) after fixing the GraphTooBig exceptions, but still faster that our current implementation. It seems that @TruffleBoundaries play a huge role in performance, so after I finish all the optimizations, I will revisit them one more time

**Comment by Alexzerntev** [comment]
retest this please

File: snapi-truffle/src/main/java/raw/runtime/truffle/ast/expressions/iterable/list/ListTakeNode.java [hunk 1]
```diff
@@ -24,123 +22,11 @@
...
-  protected ObjectList doObject(Object list, long num, @CachedLibrary("list") ListLibrary lists) {
-    Object[] innerList = (Object[]) lists.getInnerList(list);
-    if (num >= innerList.length) {
-      return (ObjectList) list;
-    }
-    Object[] result = new Object[(int) num];
-    System.arraycopy(innerList, 0, result, 0, result.length);
-    return new ObjectList(result);
+  @Specialization
+  protected Object doByte(
```
Comment (Alexzerntev) [inline snapi-truffle/src/main/java/raw/runtime/truffle/ast/expressions/iterable/list/ListTakeNode.java:28]: rename

**Review by bgaidioz — APPROVED** [review]
We reviewed it together.

**Pull request merged**